File: julia/abc200_209/abc200_E.jl

using Random
infile = stdin
## Type Shortcuts (to save my wrists and fingers :))
const I = Int64; const VI = Vector{I}; const SI = Set{I}; const PI = NTuple{2,I};
const TI = NTuple{3,I}; const QI = NTuple{4,I}; const VPI = Vector{PI}; const SPI = Set{PI}
const VC = Vector{Char}; const VS = Vector{String}; VB = Vector{Bool}; VVI = Vector{Vector{Int64}}
const F = Float64; const VF = Vector{F}; const PF = NTuple{2,F}

gs()::String = rstrip(readline(infile))
gi()::Int64 = parse(Int64, gs())
gf()::Float64 = parse(Float64,gs())
gss()::Vector{String} = split(gs())
gis()::Vector{Int64} = [parse(Int64,x) for x in gss()]
gfs()::Vector{Float64} = [parse(Float64,x) for x in gss()]

function solve(N::I,K::I)::String
    dp::Array{I,2} = fill(0,3,3N)
    ## Do first level explicitly
    for i in 1:N; dp[1,i] = 1; end
    ## Do second & third level
    deriv::VI = fill(0,3N+1)
    for l in (2,3)
        fill!(deriv,0)
        for i in 1:(l==2 ? N : 2N)
            deriv[i+1] += dp[l-1,i]
            deriv[i+N+1] -= dp[l-1,i]
        end
        cur::I = 0
        for i in 1:(l == 2 ? 2N : 3N)
            cur += deriv[i]
            dp[l,i] = cur
        end
    end


    ## Find the sum
    cum = 0; mysum = 0
    for i in 3:3N; 
        newcum::I = cum + dp[3,i]
        if newcum >= K; K -= cum; mysum = i; break; else; cum = newcum; end
    end
    ## Find the first element
    cum = 0; firstdig = 0
    for i in 1:N
        newcum::I = cum + dp[2,mysum-i]
        if newcum >= K; K -= cum; firstdig = i; break; else; cum = newcum; end
    end
    ## Find the second element
    cum = 0; seconddig = 0
    for i in 1:N
        newcum::I = cum + dp[1,mysum-firstdig-i]
        if newcum >= K; K -= cum; seconddig = i; break; else; cum = newcum; end
    end
    thirddig = mysum-firstdig-seconddig
    return "$firstdig $seconddig $thirddig"
end 

function main(infn="")
    global infile
    infile = (infn != "") ? open(infn,"r") : length(ARGS) > 0 ? open(ARGS[1],"r") : stdin
    N,K = gis()
    ans = solve(N,K)
    println(ans)
end

Random.seed!(8675309)
main()

#using Profile, StatProfilerHTML
#Profile.clear()
#@profile main("B.in")
#Profile.clear()
#@profilehtml main("B.in")

File: python/abc200_209/abc200_E.py

import sys
infile = sys.stdin.buffer

def gs()  : return infile.readline().rstrip()
def gi()  : return int(gs())
def gf()  : return float(gs())
def gss() : return gs().split()
def gis() : return [int(x) for x in gss()]
def gfs() : return [float(x) for x in gss()]

def main(infn="") :
    global infile
    infile = open(infn,"r") if infn else open(sys.argv[1],"r") if len(sys.argv) > 1 else sys.stdin
    N,K = gis()
    dp = [0] * 4
    dp[1] = [0] * (N+1)
    dp[2] = [0] * (2*N+1)
    dp[3] = [0] * (3*N+1)
    cdp = [0] * 4
    cdp[1] = [0] * (N+1)
    cdp[2] = [0] * (2*N+1)
    cdp[3] = [0] * (3*N+1)
    ## Do the base case
    d1 = dp[1]; c1 = cdp[1]
    for i in range(1,N+1) : d1[i] = 1; c1[i] = c1[i-1]+1
    ## Now do the second case
    d2 = dp[2]; c2 = cdp[2]
    for i in range(2,2*N+1) :
        maxv,minv = min(i-1,N),max(i-N,1)
        d2[i] = c1[maxv]-c1[minv-1]
        c2[i] = c2[i-1] + d2[i]
    ## Now for the third case
    d3 = dp[3]; c3 = cdp[3]
    for i in range(3,3*N+1) :
        maxv,minv = min(i-1,2*N),max(i-N,2)
        d3[i] = c2[maxv]-c2[minv-1]
        c3[i] = c3[i-1] + d3[i]

    sum=3; beauty=1; taste=1
    remaining = K
    while True :
        ways = d3[sum]
        if remaining <= ways : break
        sum += 1; remaining -= ways
    while True :
        twosum = sum-beauty
        ways = 0 if twosum < 2 or twosum > 2*N else d2[twosum]
        if remaining <= ways : break
        beauty += 1; remaining -= ways
    while True :
        onesum = sum-beauty-taste
        ways = 0 if onesum < 1 or onesum > N else d1[onesum]
        if remaining <= ways : break
        taste += 1; remaining -= ways
    print(f"{beauty} {taste} {sum-beauty-taste}")

if __name__ == '__main__' :
    main()
    sys.stdout.flush()

